### PR TITLE
Monorepo support

### PR DESCRIPTION
**Pull request by john-tipper**
This is a reworked version of the initial work done in #291 by @sradi and his PR #292.  To be properly useful, this should make use of the changes in PR #307.

Note that we want to enable the user to exclude changes in submodules from being considered when calculating the version of the parent project.  When determining whether a commit takes place in a submodule, we use JGit's `addPath` command to only look at commits in a particular subdirectory.  However, JGit does not (yet) support excluding directories from its [LogCommand](https://download.eclipse.org/jgit/site/5.5.1.201910021850-r/apidocs/org/eclipse/jgit/api/LogCommand.html).  This is discussed in a StackOverflow question I asked [here](https://stackoverflow.com/questions/59143934/java-jgit-how-to-get-git-commits-not-affecting-certain-directories).

I have created a PR to JGit (https://git.eclipse.org/r/#/c/153673/) where I add a `excludePath()` method to `LogCommand`, which will allow us to exclude directories from consideration when determining whether to increment a project's version.  When this PR is merged and there is a new version released, we should be able to bump the JGit version and the code [in this section](https://github.com/john-tipper/axion-release-plugin/blob/monorepo-support/src/main/groovy/pl/allegro/tech/build/axion/release/infrastructure/git/GitRepository.groovy#L187-L195) should work (subject to the ignored tests in [MultiModuleProjectIntegrationTest](https://github.com/john-tipper/axion-release-plugin/blob/monorepo-support/src/integration/groovy/pl/allegro/tech/build/axion/release/MultiModuleProjectIntegrationTest.groovy)).  Adding the new version of JGit will require the bump to `grgit-core` to support Jgit v5 that you see in `build.gradle`.

### CODE REVIEW

**Comment by john-tipper** [comment]
Update: my PR (https://git.eclipse.org/r/#/c/153673/) was merged into JGit last night.  It should feature in the 5.6 release, not sure on a timetable for that yet.

**Comment by john-tipper** [comment]
Further update on PR https://git.eclipse.org/r/#/c/153673/ below from one of the project's maintainers:

> final release build for 5.6 is next Tuesday or Wednesday. It will be officially published with new Eclipse simultaneous release 2019-12 on Dec 18

**Comment by john-tipper** [comment]
5.6 artifact will be available on 18 Dec, so I will attempt to progress the PR at that point.
https://projects.eclipse.org/projects/technology.jgit/releases/5.6.0/plan

**Comment by john-tipper** [comment]
The new version of JGit has been incorporated, adding support for excluding directories when calculating changes, plus a new version of GrGit, which was needed to support the new JGit version.

I have removed the "Draft" status of the PR, as I think this is ready to be looked at with a view to merging; happy to accept any/all feedback.

**Comment by adamdubiel** [comment]
Thanks, I will take a look at this :)

File: src/main/groovy/pl/allegro/tech/build/axion/release/domain/VersionConfig.groovy [hunk 1]
```diff
@@ -49,6 +49,8 @@ class VersionConfig {
 
     HooksConfig hooks = new HooksConfig()
 
+    List<String> foldersToExclude = []
```
Comment (adamdubiel) [inline src/main/groovy/pl/allegro/tech/build/axion/release/domain/VersionConfig.groovy]: i wonder if we should use "directory" instead of folder (`directory` is already used in `RepositoryConfig`), so maybe `dirsToExclude`?

File: src/main/groovy/pl/allegro/tech/build/axion/release/infrastructure/di/Context.groovy [hunk 2]
```diff
@@ -31,7 +31,20 @@ class Context {
 
         instances[ScmRepository] = scmRepository
         instances[VersionService] = new VersionService(new VersionResolver(scmRepository))
+    }
+
+    public Context(Properties rules, ScmRepository scmRepository, ScmProperties scmProperties, File projectRoot, LocalOnlyResolver localOnlyResolver, VersionConfig versionConfig) {
```
Comment (adamdubiel) [inline src/main/groovy/pl/allegro/tech/build/axion/release/infrastructure/di/Context.groovy]: the idea behind having `Context` and `GradleAwareContext` classes separate was to isolate any Gradle-dependent code from plugin domain. The isolation is not perfect, but i would like to try and keep it this way. `VersionConfig` object in my view is part of Gradle config (it is mutable etc), and so it should not be imported by `Context` class. Can you change it? I know it will make constructor uglier, so maybe some `Properties` object is needed.

Comment (adamdubiel) [reply src/main/groovy/pl/allegro/tech/build/axion/release/infrastructure/di/Context.groovy]: also is old constructor needed? why not keep only one, extended constructor?

File: src/main/groovy/pl/allegro/tech/build/axion/release/domain/VersionResolver.groovy [hunk 3]
```diff
@@ -24,16 +24,33 @@ class VersionResolver {
     private final ScmRepository repository
 
     private final VersionSorter sorter
+    private ScmPosition latestChangePosition
+    private String projectRootRelativePath
+    private List<String> foldersToExcludeWhenResolvingLatestRelevantRootProjectCommit
 
     VersionResolver(ScmRepository repository) {
+        this(repository, null, Collections.emptyList())
```
Comment (adamdubiel) [inline src/main/groovy/pl/allegro/tech/build/axion/release/domain/VersionResolver.groovy]: if we need only one constructor - go ahead and remove this one :)

File: src/main/groovy/pl/allegro/tech/build/axion/release/domain/VersionResolver.groovy [hunk 4]
```diff
@@ -24,16 +24,33 @@ class VersionResolver {
...
+    VersionResolver(ScmRepository repository, String projectRootRelativePath, List<String> excludeFolders) {
         this.repository = repository
+        this.projectRootRelativePath = projectRootRelativePath
         this.sorter = new VersionSorter()
+        this.foldersToExcludeWhenResolvingLatestRelevantRootProjectCommit = excludeFolders
     }
 
     VersionContext resolveVersion(VersionProperties versionRules, TagProperties tagProperties, NextVersionProperties nextVersionProperties) {
-        ScmPosition position = repository.currentPosition()
+        if (projectRootRelativePath != null) {
```
Comment (adamdubiel) [inline src/main/groovy/pl/allegro/tech/build/axion/release/domain/VersionResolver.groovy]: from what i see `projectRootRelativePath` is never `null` on production (might be null in tests) - why keep this switch? it makes prod code more complicated and i don't want tests to complicate prod code

File: build.gradle [hunk 5]
```diff
@@ -16,6 +16,13 @@ if (project.hasProperty('mavenCentral')) {
     apply from: 'gradle/mavenCentral.gradle'
 }
 
+sourceSets {
```
Comment (adamdubiel) [inline build.gradle]: is this change necessary?

File: docs/configuration/basic_usage.md [hunk 6]
```diff
@@ -117,6 +117,26 @@ my-service
 my-service-client
 ```
 
+Use the `foldersToExclude` configuration parameter to identify submodules
```
Comment (adamdubiel) [inline docs/configuration/basic_usage.md]: I think this prominent feature should be also more visible in configuration. In general the long-term maintenance of this would be easier if there was:

* separate configuration object (features might pile up in the future)
* feature switch to allow using old flow without `null` or `isEmpty` checks

I suggest creating `MonorepoConfig` (transformed into `MonorepoProperties` to keep Gradle/domain isolation):

```
class MonorepoConfig {
    boolean enabled = false;
    List<String> projectDirs = [];
}

class VersionConfig {
    MonorepoConfig monorepo;
}
```

File: src/main/groovy/pl/allegro/tech/build/axion/release/domain/VersionResolver.groovy (hunk 4 above)
Comment (adamdubiel) [reply src/main/groovy/pl/allegro/tech/build/axion/release/domain/VersionResolver.groovy]: also: this is the place where old vs new flow can be distinguished using feature switch i proposed :)

File: src/main/groovy/pl/allegro/tech/build/axion/release/domain/VersionResolver.groovy [hunk 7]
```diff
@@ -64,23 +81,22 @@ class VersionResolver {
         boolean forceSnapshot = versionProperties.forceSnapshot
 
         Map currentVersionInfo, previousVersionInfo
-        TagsOnCommit latestTags = repository.latestTags(releaseTagPattern)
-        currentVersionInfo = versionFromTaggedCommits([latestTags], false, nextVersionTagPattern,
+        TaggedCommits latestTaggedCommit = TaggedCommits.fromLatestCommit(repository, releaseTagPattern, latestChangePosition)
+        currentVersionInfo = versionFromTaggedCommits(latestTaggedCommit, false, nextVersionTagPattern,
             versionFactory, forceSnapshot)
+        boolean onCommitWithLatestChange = currentVersionInfo.commit == latestChangePosition.revision
+        println("currentVersion = ${currentVersionInfo.commit}, latestChangePosition=${latestChangePosition.revision}, onCommitWithLatestChange=${onCommitWithLatestChange}, currentVersionInfo.isNextVersion=${currentVersionInfo.isNextVersion}")
```
Comment (adamdubiel) [inline src/main/groovy/pl/allegro/tech/build/axion/release/domain/VersionResolver.groovy]: should it stay here? :)

**Comment by john-tipper** [comment]
Thanks very much, let me dig into these get back to you with changes asap.

File: src/main/groovy/pl/allegro/tech/build/axion/release/domain/VersionConfig.groovy (hunk 1 above)
Comment (john-tipper) [reply src/main/groovy/pl/allegro/tech/build/axion/release/domain/VersionConfig.groovy]: Agreed.

File: build.gradle (hunk 5 above)
Comment (john-tipper) [reply build.gradle]: Yes, didn't see the second block when I added this block, so have merged with the `sourceSets` block further down in the file.

File: src/main/groovy/pl/allegro/tech/build/axion/release/domain/VersionResolver.groovy (hunk 7 above)
Comment (john-tipper) [reply src/main/groovy/pl/allegro/tech/build/axion/release/domain/VersionResolver.groovy]: Sorry, I blame a tired programmer working late, personally :-)  I've removed it.

File: src/main/groovy/pl/allegro/tech/build/axion/release/infrastructure/di/Context.groovy (hunk 2 above)
Comment (john-tipper) [reply src/main/groovy/pl/allegro/tech/build/axion/release/infrastructure/di/Context.groovy]: Please would you mind confirming your view on what `Context` should do - it's a class to encapsulate all the Git-specific classes/tools etc, right?  So `Context` should have some sort of properties (`MonoRepoProperties`) that are used to process mono-repos, and this class would be created via a similar static method to the call to `LocalOnlyResolverFactory.create(...)`?

e.g. 
`MonoRepoProperties.createFromConfig(MonoRepoConfig config)
`
where `MonoRepoconfig` is a class contained within `VersionConfig`?

Comment (adamdubiel) [reply src/main/groovy/pl/allegro/tech/build/axion/release/infrastructure/di/Context.groovy]: `Context` is a class to build the whole core environment - meaning version reading & resolution. Core logic should have nothing to do with Gradle.

Each `*Config` file is transformed into `*Properties` using a separate `*Factory`. Factory transforms Gradle domain objects (`*Config`) into core domain (`*Properties`) while handing cmd args and env overwrites. `*Factory` classes belong to Gradle, as they are aware of Gradle.

So, in general, the second answer :)

Comment (john-tipper) [reply src/main/groovy/pl/allegro/tech/build/axion/release/infrastructure/di/Context.groovy]: Amended.

File: src/main/groovy/pl/allegro/tech/build/axion/release/domain/VersionResolver.groovy (hunk 3 above)
Comment (john-tipper) [reply src/main/groovy/pl/allegro/tech/build/axion/release/domain/VersionResolver.groovy]: Amended.

File: src/main/groovy/pl/allegro/tech/build/axion/release/domain/VersionResolver.groovy (hunk 4 above)
Comment (john-tipper) [reply src/main/groovy/pl/allegro/tech/build/axion/release/domain/VersionResolver.groovy]: Yes the check was to see if the path was `null` as a result of tests, which it no longer can be due to the changes of constructor.  However, all resolution of versions must go through the new method, as the project may be running as a submodule, i.e. `:submodule:currentVersion`, as opposed to `:currentversion`.  The configuration parsed by that submodule is its own config and not that of the parent project, which is where the feature flag to enable monorepo support would be set (i.e. the subprojects would also need to have the feature flag set to true in their own config). I don't think it's correct that submodules should have to also have the flag set. This means that there's no need for a feature flag.

**Comment by john-tipper** [comment]
I've made some more changes, so hopefully it's closer to what you'd like.  The only thing I didn't implement is the feature toggle - this is because even if a project does not have a enable flag set, that project may well be running as a submodule of a parent project which does have the flag set.

File: src/main/groovy/pl/allegro/tech/build/axion/release/domain/properties/VersionProperties.groovy [hunk 8]
```diff
@@ -17,9 +17,11 @@ class VersionProperties {
     final Closure<Version> versionIncrementer
 
     final boolean sanitizeVersion
-    
+
     final boolean useHighestVersion
 
+    final MonorepoProperties monorepoProperties
```
Comment (adamdubiel) [inline src/main/groovy/pl/allegro/tech/build/axion/release/domain/properties/VersionProperties.groovy]: i think `MonorepoProperties` deserves to be a top-level `Properties` object and doesn't have to be nested in `VersionProperties`, but since this is internal object, it can be changed at any time :)

**Comment by adamdubiel** [comment]
You are right about feature flag, i didn't take this into a count. Everything looks good to me, added one minor thing. If you feel like changing it it would be great. If not - I'm willing to merge it anyways. Lots of good work and great tests in here :) This will be shipped as 1.11.

File: src/main/groovy/pl/allegro/tech/build/axion/release/domain/properties/VersionProperties.groovy (hunk 8 above)
Comment (john-tipper) [reply src/main/groovy/pl/allegro/tech/build/axion/release/domain/properties/VersionProperties.groovy]: `VersionService` currently receives in `currentVersion()` and `currentDecoratedVersion()` a `VersionProperties` object.  To me, this seems quite clean, in that `VersionService` only receives properties that are related to itself and its behaviour is not changed by properties outside of its concern (e.g. it doesn't receive a `HooksProperties` object, for instance).  `VersionService` in turn passes these properties into `VersionResolver`.

If we were to move `MonorepoProperties` out of `VersionProperties` and into `Properties` then `VersionService` would need to have passed to it either the full `Properties` object or both  `MonorepoProperties` and `VersionProperties`.  I'm happy to make the change if you'd like, but it seems to me to result in somewhat more unwieldy service classes - what do you think?

Comment (adamdubiel) [reply src/main/groovy/pl/allegro/tech/build/axion/release/domain/properties/VersionProperties.groovy]: I feel convinced. Let's keep it as it is :)

Comment (john-tipper) [reply src/main/groovy/pl/allegro/tech/build/axion/release/domain/properties/VersionProperties.groovy]: Thanks :-)

**Comment by john-tipper** [comment]
Thanks very much for the code reviews, it was fun getting the design to work nicely!  When do you expect to push the 1.11.0 release?

**Comment by adamdubiel** [comment]
Probably on Friday or during the weekend :) I want to play around with monorepo on snapshot builds.

**Comment by john-tipper** [comment]
Thanks for merging, I've updated the docs to refer to the new name for the monorepo configuration block in a separate PR #318 (`foldersToExclude` was changed to `projectDirs`).  Sorry, I missed this in this PR.
